### PR TITLE
adding kube-apiserver starting option tests

### PR DESCRIPTION
**Pull request by huangjiuyuan**
**What this PR does / why we need it**:
There is only one test for `--enable-swagger-ui` option in `cmd/kube-apiserver/app/options/options_test.go`, I have expanded this test with more kube-apiserver starting options.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

`NONE`

### CODE REVIEW

**Comment by mikedanese** [comment]
What does this bugs would this test actually catch? Why is this not sufficiently covered by a unit test in the flags library?

**Comment by huangjiuyuan** [comment]
HI, @mikedanese 
Thanks for your feedback. Could you please be more specific about the location of the flags library? These tests ensure that the values of flags can be passed correctly to the apiserver. By the way, the comment says `TODO: This only tests the enable-swagger-ui flag for now. Expand the test to include other flags as well.` on the above, so I expand the test and include more flags in this unit test.

**Comment by huangjiuyuan** [comment]
@sttts Could you please take a quick look at this, Thank you!

File: cmd/kube-apiserver/app/options/options_test.go [hunk 1]
```diff
@@ -28,15 +28,217 @@ func TestAddFlagsFlag(t *testing.T) {
 	f := pflag.NewFlagSet("addflagstest", pflag.ContinueOnError)
 	s := NewServerRunOptions()
 	s.AddFlags(f)
+	if s.GenericServerRunOptions.AdvertiseAddress != nil {
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: Not sure the upper block here adds anything.

File: cmd/kube-apiserver/app/options/options_test.go [hunk 2]
```diff
@@ -28,15 +28,217 @@ func TestAddFlagsFlag(t *testing.T) {
...
+		"--cloud-provider=daocloud",
+		"--cors-allowed-origins=10.10.10.100,10.10.10.200",
+		"--enable-aggregator-routing=true",
+		"--enable-logs-handler=false",
 		"--enable-swagger-ui=true",
 	}
 	f.Parse(args)
+
+	// Check option --advertise-address
+	if s.GenericServerRunOptions.AdvertiseAddress.String() != "192.168.10.10" {
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: How about replacing all the single if-then clauses with a DeepEqual and a big struct? Would be much easier to  maintain.

**Comment by huangjiuyuan** [comment]
Hi @sttts 
Thanks for your advice. I have replaced all the single if-then clauses with a DeepEqual and a big struct for convenience on maintenance.

File: cmd/kube-apiserver/app/options/options_test.go [hunk 3]
```diff
@@ -17,26 +17,164 @@ limitations under the License.
...
+	parsedOps.ClientCAFile = s.Authentication.ClientCert.ClientCA
+	parsedOps.CloudConfig = s.CloudProvider.CloudConfigFile
+	parsedOps.CloudProvider = s.CloudProvider.CloudProvider
+	parsedOps.CorsAllowedOrigins = s.GenericServerRunOptions.CorsAllowedOriginList
+	parsedOps.EnableAggregatorRouting = s.EnableAggregatorRouting
+	parsedOps.EnableLogsHandler = s.EnableLogsHandler
+	parsedOps.EnableSwaggerUI = s.Features.EnableSwaggerUI
+
+	// Values which parsed values should be equal to
+	testOps := Options{}
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: Why do we need this special type? Why can't we just use the existing `Options` type, create the instance of that with the expected values and compare `s` with that?

Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: Compare what we do in OpenShift: https://github.com/openshift/origin/blob/master/pkg/cmd/server/kubernetes/master/master_config_test.go#L69

**Comment by huangjiuyuan** [comment]
@sttts 
Thank you for your help. I have created the instance of `Options` with the expected values and compared the parsed values with expected values.

File: cmd/kube-apiserver/app/options/options_test.go [hunk 4]
```diff
@@ -17,26 +17,167 @@ limitations under the License.
...
+		s.CloudProvider.CloudConfigFile,
+		s.CloudProvider.CloudProvider,
+		s.GenericServerRunOptions.CorsAllowedOriginList,
+		s.EnableAggregatorRouting,
+		s.EnableLogsHandler,
+		s.Features.EnableSwaggerUI,
+	}
+
+	// Expected apiserver run options set by args
+	expected := Options{
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: What about converting this `Options` instance in the format of `s`? Then we don't need our local `Options` type. Makes much easier to maintain.

**Comment by huangjiuyuan** [comment]
@sttts
I have converted the `Options` instance in the format of `ServerRunOptions`, which could be easier to maintain.

File: cmd/kube-apiserver/app/options/options_test.go [hunk 5]
```diff
@@ -17,26 +17,179 @@ limitations under the License.
...
+		},
+		APIEnablement: &kubeoptions.APIEnablementOptions{
+			RuntimeConfig: utilconfig.ConfigurationMap{},
+		},
+		EnableLogsHandler:       false,
+		EnableAggregatorRouting: true,
+	}
+
+	if !reflect.DeepEqual(expected, s) {
+		t.Logf("expected %s, got %s", expected, s)
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: why not combine the log and the error?

**Comment by sttts** [comment]
Well done! One nit is left. Then it looks good.

**Comment by sttts** [comment]
@huangjiuyuan you have to run hack/update-bazel.sh.

**Comment by huangjiuyuan** [comment]
Hi, @sttts 
Thank you for your patience. Could you please take a quick review on this?

**Comment by sttts** [comment]
@huangjiuyuan looks nearly ready to go. Last thing: we have `ObjectDiff` at `vendor/k8s.io/apimachinery/pkg/util/diff/diff.go:56` to print out the difference of expected and gotten options. If that does not work, at least print out both strings to have an easy chance to see the difference.

**Comment by huangjiuyuan** [comment]
@sttts `ObjectDiff` does not work because there are some functions in the struct. However, I find that I can use `ObjectReflectDiff` to print out the difference and It works fine with every fields in the `ServerRunOptions` struct.

**Comment by sttts** [comment]
> I find that I can use ObjectReflectDiff to print out the difference

cool!

File: cmd/kube-apiserver/app/options/options_test.go [hunk 6]
```diff
@@ -17,26 +17,179 @@ limitations under the License.
 package options
 
 import (
+	"net"
+	"reflect"
 	"testing"
+	"time"
 
 	"github.com/spf13/pflag"
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options_test.go]: last little nit: split split the non-kube import with an empty line

Comment (huangjiuyuan) [inline cmd/kube-apiserver/app/options/options_test.go]: Hi, @sttts 
I have added an empty line to split non-kube import and tube import. Is it ready to go?

**Comment by sttts** [comment]
/lgtm

🚀

**Comment by huangjiuyuan** [comment]
@sttts 
Thank you very much!

**Pull request merged**